{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b7f7cf58-11cc-4908-8eba-78a68b003c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e2b92c7-3556-4529-8f2a-b10f1ad749e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "coller_optimality_index = {\n",
    "    \"GCC\":51,\n",
    "    \"CUG\":5,#\n",
    "    \"GGC\":35,\n",
    "    \"CGG\":15,#\n",
    "    \"CGC\":20,#\n",
    "    \"CCC\":24,\n",
    "    \"GUG\":22,\n",
    "    \"GCG\":21,\n",
    "    \"CCG\":4,#\n",
    "    \"GAC\":40,\n",
    "    \"CUC\":10,#\n",
    "    \"UCG\":16,#\n",
    "    \"AGC\":9,#\n",
    "    \"ACC\":41,#/\n",
    "    \"ACG\":6,\n",
    "    \"GGG\":23,\n",
    "    \"GUC\":54,\n",
    "    \"AUC\":40,#/\n",
    "    \"UCC\":47,\n",
    "    \"UUC\":46,\n",
    "    \"UGC\":28,\n",
    "    \"UAC\":47,\n",
    "    \"GAG\":18,#\n",
    "    \"CAC\":42,\n",
    "    \"CAG\":11,#/\n",
    "    \"UGG\":34,\n",
    "    \"AAC\":41, #last sub-optimal codon\n",
    "    \"AAG\":48,\n",
    "    \"CGU\":44,\n",
    "    \"GCU\":56,\n",
    "    \"CGA\":1,#\n",
    "    \"AUG\":30,\n",
    "    \"AGG\":2,#\n",
    "    \"CCU\":36,\n",
    "    \"GGU\":55,\n",
    "    \"CCA\":50,\n",
    "    \"GCA\":26,\n",
    "    \"UUG\":53,\n",
    "    \"CUA\":12,#\n",
    "    \"GUA\":3,#\n",
    "    \"CUU\":13,#\n",
    "    \"UCU\":48,#/\n",
    "    \"GUU\":52,\n",
    "    \"UGU\":33,\n",
    "    \"UUU\":29,\n",
    "    \"GAU\":31,\n",
    "    \"AGU\":7,\n",
    "    \"ACA\":11,#/\n",
    "    \"GGA\":17,#\n",
    "    \"UAU\":25,#\n",
    "    \"UUA\":32,\n",
    "    \"UCA\":19,\n",
    "    \"CAU\":27,\n",
    "    \"ACU\":49,#/\n",
    "    \"AUU\":38,\n",
    "    \"AUA\":0,#\n",
    "    \"AAU\":14,#\n",
    "    \"CAA\":43,\n",
    "    \"AGA\":37,\n",
    "    \"GAA\":45,\n",
    "    \"AAA\":8 #    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "14f87ff6-b778-449a-86a0-deee6ac8b3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_optimality_index = {\n",
    "    \"GCC\":0,\n",
    "    \"CUG\":1,\n",
    "    \"GGC\":2,\n",
    "    \"CGG\":3,\n",
    "    \"CGC\":4,\n",
    "    \"CCC\":5,\n",
    "    \"GUG\":6,\n",
    "    \"GCG\":7,\n",
    "    \"CCG\":8,\n",
    "    \"GAC\":9,\n",
    "    \"CUC\":10,\n",
    "    \"UCG\":11,\n",
    "    \"AGC\":12,\n",
    "    \"ACC\":13,\n",
    "    \"ACG\":14,\n",
    "    \"GGG\":15,\n",
    "    \"GUC\":16,\n",
    "    \"AUC\":17,\n",
    "    \"UCC\":18,\n",
    "    \"UUC\":19,\n",
    "    \"UGC\":20,\n",
    "    \"UAC\":21,\n",
    "    \"GAG\":22,\n",
    "    \"CAC\":23,\n",
    "    \"CAG\":24,\n",
    "    \"UGG\":25,\n",
    "    \"AAC\":26, #last sub-optimal codon\n",
    "    \"AAG\":27,\n",
    "    \"CGU\":28,\n",
    "    \"GCU\":29,\n",
    "    \"CGA\":30,\n",
    "    \"AUG\":31,\n",
    "    \"AGG\":32,\n",
    "    \"CCU\":33,\n",
    "    \"GGU\":34,\n",
    "    \"CCA\":35,\n",
    "    \"GCA\":36,\n",
    "    \"UUG\":37,\n",
    "    \"CUA\":38,\n",
    "    \"GUA\":39,\n",
    "    \"CUU\":40,\n",
    "    \"UCU\":41,\n",
    "    \"GUU\":42,\n",
    "    \"UGU\":43,\n",
    "    \"UUU\":44,\n",
    "    \"GAU\":45,\n",
    "    \"AGU\":46,\n",
    "    \"ACA\":47,\n",
    "    \"GGA\":48,\n",
    "    \"UAU\":49,\n",
    "    \"UUA\":50,\n",
    "    \"UCA\":51,\n",
    "    \"CAU\":52,\n",
    "    \"ACU\":53,\n",
    "    \"AUU\":54,\n",
    "    \"AUA\":55,\n",
    "    \"AAU\":56,\n",
    "    \"CAA\":57,\n",
    "    \"AGA\":58,\n",
    "    \"GAA\":59,\n",
    "    \"AAA\":60    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60d04fbe-ef32-46c8-b479-b42cda0bbef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matching_positions = [x for (x, y) in enumerate(zip(codon_optimality_index.values(), coller_optimality_index.values())) if x == y]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4dd8b086-4aa4-482c-b489-7c5f5ed30c1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({41: 2,\n",
       "         10: 2,\n",
       "         40: 2,\n",
       "         42: 2,\n",
       "         48: 2,\n",
       "         11: 2,\n",
       "         49: 2,\n",
       "         51: 1,\n",
       "         6: 1,\n",
       "         37: 1,\n",
       "         16: 1,\n",
       "         22: 1,\n",
       "         26: 1,\n",
       "         24: 1,\n",
       "         23: 1,\n",
       "         5: 1,\n",
       "         17: 1,\n",
       "         7: 1,\n",
       "         25: 1,\n",
       "         54: 1,\n",
       "         47: 1,\n",
       "         30: 1,\n",
       "         20: 1,\n",
       "         43: 1,\n",
       "         36: 1,\n",
       "         45: 1,\n",
       "         56: 1,\n",
       "         1: 1,\n",
       "         32: 1,\n",
       "         3: 1,\n",
       "         38: 1,\n",
       "         55: 1,\n",
       "         50: 1,\n",
       "         28: 1,\n",
       "         53: 1,\n",
       "         12: 1,\n",
       "         4: 1,\n",
       "         13: 1,\n",
       "         52: 1,\n",
       "         35: 1,\n",
       "         31: 1,\n",
       "         33: 1,\n",
       "         8: 1,\n",
       "         18: 1,\n",
       "         27: 1,\n",
       "         34: 1,\n",
       "         21: 1,\n",
       "         29: 1,\n",
       "         0: 1,\n",
       "         15: 1,\n",
       "         44: 1,\n",
       "         39: 1,\n",
       "         46: 1,\n",
       "         9: 1})"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(coller_optimality_index.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad8a1225-fe80-4ad5-8483-ae3ed5ef3854",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_aa_codon_optimality_index = {'UUU':1,#Phe\n",
    "                                 'UUC':19/44,\n",
    "                                 'UUA':1,#Leu\n",
    "                                 'UUG':0.74,\n",
    "                                 'CUU':0.8,\n",
    "                                 'CUC':0.2,\n",
    "                                 'CUA':0.76,\n",
    "                                 'CUG':0.04,\n",
    "                                 'AUU':54/55,#Ile\n",
    "                                 'AUC':17/55,\n",
    "                                 'AUA':1,\n",
    "                                 'AUG':1,#Met\n",
    "                                 'GUU':1,#Val\n",
    "                                 'GUC':16/42,\n",
    "                                 'GUA':39/42,\n",
    "                                 'GUG':6/42,\n",
    "                                 'UCU':41/51, #Ser\n",
    "                                 'UCC':18/51,\n",
    "                                 'UCA':1,\n",
    "                                 'UCG':11/51,\n",
    "                                 'CCU':33/35,#Pro\n",
    "                                 'CCC':5/35,\n",
    "                                 'CCA':1,\n",
    "                                 'CCG':8/35,\n",
    "                                 'ACU':1, #Thr\n",
    "                                 'ACC':13/53,\n",
    "                                 'ACA':47/53,\n",
    "                                 'ACG':14/53,\n",
    "                                 'GCU':29/36,#Ala\n",
    "                                 'GCC':0,\n",
    "                                 'GCA':1,\n",
    "                                 'GCG':7/36,\n",
    "                                 'UAU':1,#Tyr\n",
    "                                 'UAC':21/49,\n",
    "                                 'CAU':1, #His\n",
    "                                 'CAC':23/52,\n",
    "                                 'CAA':1, #Gln\n",
    "                                 'CAG':24/57,\n",
    "                                 'AAU':1, #Asn\n",
    "                                 'AAC':26/56,\n",
    "                                 'AAA':1, #Lys\n",
    "                                 'AAG':27/60,\n",
    "                                 'GAU':1, #Asp\n",
    "                                 'GAC':9/45,\n",
    "                                 'GAA':1, #Glu\n",
    "                                 'GAG':22/59,\n",
    "                                 'UGU':1, #Cys\n",
    "                                 'UGC':20/43,\n",
    "                                 'UGG':1, #Trp\n",
    "                                 'CGU':28/58, #Arg\n",
    "                                 'CGC':4/58,\n",
    "                                 'CGA':30/58,\n",
    "                                 'CGG':3/58,\n",
    "                                 'AGA':1,\n",
    "                                 'AGG':32/58,\n",
    "                                 'AGU':1, #Ser\n",
    "                                 'AGC':12/46,\n",
    "                                 'GGU':34/48, #Gly\n",
    "                                 'GGC':2/48,\n",
    "                                 'GGA':1,\n",
    "                                 'GGG':15/48\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b30a413-b018-4998-ab29-a22a3a8b85cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCodonOptimality(sequence,optimality_index):\n",
    "    if sequence is not None:\n",
    "        score_list = []\n",
    "        stop_codons = ['UAG','UAA','UGA']\n",
    "        rna_sequence = sequence.replace('T','U')\n",
    "        rna_sequence = rna_sequence.replace('\\n','')\n",
    "        pre_orf = True\n",
    "        for i in range(0,len(rna_sequence)-2,3):\n",
    "            codon = rna_sequence[i:i+3]\n",
    "            if codon in stop_codons:\n",
    "                    break\n",
    "            else:\n",
    "                try:\n",
    "                    score_list.append(optimality_index[codon])\n",
    "                except KeyError:\n",
    "                    score_list.append(30)\n",
    "        codon_opt_score = np.mean(score_list)\n",
    "    else:\n",
    "        codon_opt_score = None\n",
    "    return codon_opt_score, len(score_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "19cc9723-7ed8-4e5c-8dc7-22488b3d7da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCDNA(gene_id,fail_list):\n",
    "    # Define the Ensembl REST API endpoint\n",
    "    ensembl_api_url = 'https://rest.ensembl.org/'\n",
    "\n",
    "    # Define the species (human)\n",
    "    species = 'human'\n",
    "\n",
    "    # Define the endpoint URL to retrieve the top transcript ID for the specified gene\n",
    "    transcript_url = f\"{ensembl_api_url}lookup/id/{gene_id}?expand=1;species={species}\"\n",
    "\n",
    "    # Send GET request to retrieve information about the gene\n",
    "    response_transcript = requests.get(transcript_url, headers={\"Content-Type\": \"application/json\"})\n",
    "    \n",
    "    if response_transcript.status_code == 200:\n",
    "        # Extract the top transcript ID from the response\n",
    "        top_transcript_id = response_transcript.json()['Transcript'][0]['id']\n",
    "\n",
    "        # Define the endpoint URL to retrieve cDNA sequence for the top transcript ID\n",
    "        request_url = f\"{ensembl_api_url}sequence/id/{top_transcript_id}?type=cdna;species={species}\"\n",
    "\n",
    "        # Send GET request to retrieve cDNA sequence for the top transcript ID\n",
    "        response_cDNA = requests.get(request_url, headers={\"Content-Type\": \"text/plain\"})\n",
    "\n",
    "        if response_cDNA.status_code == 200:\n",
    "            sequence = response_cDNA.text\n",
    "            with open(f'/mnt/scratchc/ghlab/toby/transcripts/{gene_id}.txt','w') as file:\n",
    "                file.write(sequence)\n",
    "            \n",
    "            #print(f\"cDNA sequence for the top transcript ID {top_transcript_id} (gene ID {gene_id}, human):\\n{cDNA_sequence}\")\n",
    "        else:\n",
    "            print(f\"Error fetching cDNA sequence: {response_cDNA.status_code}\")\n",
    "            fail_list.append(gene_id)\n",
    "    else:\n",
    "        print(f\"Error fetching transcript information: {response_transcript.status_code}\")\n",
    "        fail_list.append(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d480b77c-6d83-408b-a0b8-de086d51e57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cds_sequence(gene_id, fail_list):\n",
    "    # Define the Ensembl REST API endpoint\n",
    "    ensembl_api_url = 'https://rest.ensembl.org/'\n",
    "\n",
    "    # Define the species (human)\n",
    "    species = 'human'\n",
    "\n",
    "    # Define the endpoint URL to retrieve the top transcript ID for the specified gene\n",
    "    transcript_url = f\"{ensembl_api_url}lookup/id/{gene_id}?expand=1;species={species}\"\n",
    "\n",
    "    # Send GET request to retrieve information about the gene\n",
    "    response_transcript = requests.get(transcript_url, headers={\"Content-Type\": \"application/json\"})\n",
    "\n",
    "    if response_transcript.status_code == 200:\n",
    "        # Extract the top transcript ID from the response\n",
    "        top_transcript_id = response_transcript.json()['Transcript'][0]['id']\n",
    "\n",
    "        # Define the endpoint URL to retrieve CDS sequence for the top transcript ID\n",
    "        request_url = f\"{ensembl_api_url}sequence/id/{top_transcript_id}?type=cds;species={species}\"\n",
    "\n",
    "        # Send GET request to retrieve CDS sequence for the top transcript ID\n",
    "        response_cds = requests.get(request_url, headers={\"Content-Type\": \"text/plain\"})\n",
    "\n",
    "        if response_cds.status_code == 200:\n",
    "            cds_sequence = response_cds.text\n",
    "            with open(f'/mnt/scratchc/ghlab/toby/cds_sequences/{gene_id}.txt', 'w') as file:\n",
    "                file.write(cds_sequence)\n",
    "        else:\n",
    "            print(f\"Error fetching CDS sequence: {response_cds.status_code}\")\n",
    "            fail_list.append(gene_id)\n",
    "    else:\n",
    "        print(f\"Error fetching transcript information: {response_transcript.status_code}\")\n",
    "        fail_list.append(gene_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65f8fdb9-dd52-4900-81d0-078c459e9d9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<mask>'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_dict = []\n",
    "with (open(\"/mnt/scratchc/ghlab/toby/Geneformer/geneformer/token_dictionary.pkl\", \"rb\")) as openfile:\n",
    "    while True:\n",
    "        try:\n",
    "            token_dict.append(pickle.load(openfile))\n",
    "        except EOFError:\n",
    "            break\n",
    "\n",
    "ensg_list = list(token_dict[0].keys())\n",
    "ensg_list.pop(0)\n",
    "ensg_list.pop(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b3ec82c-c2c7-42fe-9612-5e8a3c646b94",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fail_list=[]\n",
    "for gene in ensg_list:\n",
    "    get_cds_sequence(gene,fail_list)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3b583e5-11bc-4f6a-9b43-27bc151b196f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores_list=[]\n",
    "per_aa_scores_list=[]\n",
    "present_genes=[]\n",
    "missing_genes=[]\n",
    "n_codon_list=[]\n",
    "counter = 0\n",
    "\n",
    "for ensg_id in ensg_list:\n",
    "    try:\n",
    "        with open(f'/mnt/scratchc/ghlab/toby/cds_sequences/{ensg_id}.txt') as file:\n",
    "                sequence=file.read()\n",
    "                score, n_codons = getCodonOptimality(sequence, codon_optimality_index)\n",
    "                aa_score, n_codons = getCodonOptimality(sequence, per_aa_codon_optimality_index)\n",
    "                scores_list.append(score)\n",
    "                per_aa_scores_list.append(aa_score)\n",
    "                present_genes.append(ensg_id)\n",
    "                n_codon_list.append(n_codons)\n",
    "    except FileNotFoundError:\n",
    "        missing_genes.append(ensg_id)\n",
    "    \n",
    "df = pd.DataFrame({'ensembl_id':present_genes,'codon_opt_score':scores_list,'per_aa_codon_opt_score':per_aa_scores_list,'n_codons':n_codon_list})\n",
    "df.to_csv('/mnt/scratchc/ghlab/toby/in_silico_perturb_data/cds_codon_optimalities.csv',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ede82b7-a19a-4074-95e1-984138065111",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores_list=[]\n",
    "present_genes=[]\n",
    "missing_genes=[]\n",
    "n_codons_list=[]\n",
    "\n",
    "for ensg_id in ensg_list:\n",
    "    try:\n",
    "        with open(f'/mnt/scratchc/ghlab/toby/transcripts/{ensg_id}.txt') as file:\n",
    "                sequence=file.read()\n",
    "                score, n_codons = getCodonOptimality(sequence, per_aa_codon_optimality_index)\n",
    "                scores_list.append(score)\n",
    "                n_codons_list.append(n_codons)\n",
    "                present_genes.append(ensg_id)\n",
    "    except FileNotFoundError:\n",
    "        missing_genes.append(ensg_id)\n",
    "    \n",
    "codon_opt_df = pd.read_csv('/mnt/scratchc/ghlab/toby/in_silico_perturb_data/codon_optimalities.csv')\n",
    "codon_opt_df['n_codons'] = n_codons_list\n",
    "codon_opt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85ce05c-e628-41e4-b455-14e05f0e6d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_opt_df.to_csv('/mnt/scratchc/ghlab/toby/in_silico_perturb_data/codon_optimalities.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273e35c0-7014-49c3-bd8c-a31f348964bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_opt_df[codon_opt_df['n_codons'] == 33423]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a725d161-5c17-47e2-a2ec-6e52c7d44f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(codon_opt_df['per_aa_codon_opt_score'], bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39465884-2887-42a6-92d9-113cbd5198ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/mnt/scratchc/ghlab/toby/in_silico_perturb_data/missing_genes.csv','w') as file:\n",
    "    for gene in missing_genes:\n",
    "        file.write(gene+',')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ed58fcc-1bc1-4063-90d8-2cb4a88ac713",
   "metadata": {},
   "outputs": [],
   "source": [
    "#see if genes with a higher initial stability are more/less affected.\n",
    "#PCA/UMAP of gene embeddings with codon optimality.\n",
    "#potentially average gene embeddings across cells.\n",
    "\n",
    "\n",
    "#potentially add gc content as a metric\n",
    "#boxplots of optimality category against euclidean distance.\n",
    "#perturb 2 different genes and plot distance1 against distance2\n",
    "#try to find a positive result with the model.\n",
    "#WT vs knockout scRNA-seq for 2-class classification models.\n",
    "#Consider mRNA half-lives potentially.\n",
    "\n",
    "#Could use embryonic development or stem cells vs differentiated cells.\n",
    "#Supervised UMAP with labels - codon opts and gene embeddings, cnot1 knockout/not knockout\n",
    "#Random forest to classify CNOT1 knockout vs wild-type.\n",
    "#Look at Gene Expression Omnibus to try to find an interesting dataset.\n",
    "\n",
    "#Look at heart-specific vs not active in heart TFs - boxplot heart-specific TFs vs non-heart-specific TFs/genes."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
